crates/qpy/src/formats.rs

```diff
@@ -0,0 +1,823 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::bytes::Bytes;
+use crate::expr::{read_expression, write_expression};
+use crate::params::ParameterType;
+use crate::value::{
+    BitType, CircuitInstructionType, ExpressionType, ExpressionVarDeclaration, ModifierType,
+    QPYReadData, QPYWriteData, ValueType,
+};
+use binrw::{BinRead, BinResult, BinWrite, Endian, binread, binrw, binwrite};
+use qiskit_circuit::classical::expr::Expr;
+use std::io::{Read, Seek, Write};
+use std::marker::PhantomData;
+
+/// The overall structure of the QPY file
+
+// TODO: Still need to split by versioning and have a struct for the whole file, not a single circuit
+#[derive(BinWrite, Debug)]
+#[brw(big)]
+pub struct QPYFormatV15 {
+    pub header: CircuitHeaderV12Pack,
+    pub standalone_vars: Vec<ExpressionVarDeclarationPack>,
+    pub annotation_headers: AnnotationHeaderStaticPack,
+    pub custom_instructions: CustomCircuitInstructionsPack,
+    pub instructions: Vec<CircuitInstructionV2Pack>,
+    pub calibrations: CalibrationsPack,
+    pub layout: LayoutV2Pack,
+}
+
+#[binrw]
+#[brw(big)]
+#[derive(Debug)]
+pub struct CircuitHeaderV12Pack {
+    #[bw(calc = circuit_name.len() as u16)]
+    pub name_size: u16,
+    pub global_phase_type: ValueType,
+    #[bw(calc = global_phase_data.len() as u16)]
+    pub global_phase_size: u16,
+    pub num_qubits: u32,
+    pub num_clbits: u32,
+    #[bw(calc = metadata.len() as u64)]
+    pub metadata_size: u64,
+    #[bw(calc = registers.len() as u32)]
+    pub num_registers: u32,
+    pub num_instructions: u64,
+    pub num_vars: u32,
+    #[br(parse_with = read_string, args(name_size as usize))]
+    #[bw(write_with = write_string)]
```

Comment:
Should we go with something like this: 
```suggestion
    #[br(count = name_size, try_map = String::from_utf8)]
    #[bw(map = |s| s.as_bytes())]
```
And drop `read_string` and `write_string`?